src/map/utils/battleutils.cpp

```diff
@@ -551,7 +551,56 @@ bool HandleSpikesDamage(CBattleEntity* PAttacker, CBattleEntity* PDefender, apAc
     Action->spikesMessage = 44;
     Action->spikesParam = PDefender->getMod(MOD_SPIKES_DMG);
 
-    if (Action->spikesEffect > 0)
+    // Handle Retaliation
+    if (PDefender->StatusEffectContainer->HasStatusEffect(EFFECT_RETALIATION)
+        && battleutils::GetHitRate(PDefender, PAttacker) > WELL512::irand()%100
+        && isFaceing(PAttacker->loc.p, PDefender->loc.p, 40))
+    {
+        // Retaliation rate is based on player acc vs mob evasion. Missed retaliations do not even display in log.
+        // Other theories exist were not proven or reliably tested (I have to assume to many things to even consider JP translations about weapon delay), this at least has data to back it up.
+        // https://web.archive.org/web/20141228105335/http://www.bluegartr.com/threads/120193-Retaliation-Testing?s=7a6221e10ffdfaa6a7f5e8f0387f787d&p=4620727&viewfull=1#post4620727
+        Action->reaction = REACTION_HIT;
+        Action->spikesEffect = SUBEFFECT_COUNTER;
+
+        if (battleutils::IsAbsorbByShadow(PAttacker)) // Struck a shadow
+        {
+            Action->spikesMessage = 535;
+        }
+        else // Struck the target
+        {
+            if (PDefender->objtype == TYPE_PC)
+            {
+                // Check for skillup
+                uint8 skilltype = PDefender->m_Weapons[SLOT_MAIN]->getSkillType();
+                charutils::TrySkillUP((CCharEntity*)PDefender, (SKILLTYPE)skilltype, PAttacker->GetMLevel());
+
+                // Check if crit
+                bool crit = battleutils::GetCritHitRate(PDefender, PAttacker, true) > WELL512::irand()%100;
+
+                // Dmg math.
+                float DamageRatio = GetDamageRatio(PDefender, PAttacker, crit, 0);
+                uint16 dmg = (uint32)((PDefender->GetMainWeaponDmg() + battleutils::GetFSTR(PDefender, PAttacker, SLOT_MAIN)) * DamageRatio);
+                dmg = attackutils::CheckForDamageMultiplier(((CCharEntity*)PDefender), PDefender->m_Weapons[SLOT_MAIN], dmg, ATTACK_NORMAL);
+                uint16 bonus = dmg * (PDefender->getMod(MOD_RETALIATION) /100);
+                dmg = dmg + bonus;
+
+                // FINISH HIM! dun dun dun
+                // TP and stoneskin are handled inside TakePhysicalDamage
+                Action->spikesMessage = 536;
+                Action->spikesParam = battleutils::TakePhysicalDamage(PDefender, PAttacker, dmg, false, SLOT_MAIN, 1, NULL, true);
+                PAttacker->addHP(-Action->spikesParam);
+            }
+            /*
+            else

```

Comment:
just put the skill up code in the if TYPE_PC block? pretty sure the rest of that is all fine for mobs, then you don't have to worry about it being not implemented later